KGE documentation - main terms and parameters (#660)
neo4j/graph-data-science-client: doc/modules/ROOT/pages/python-runtime/knowledge-graph-embeddings.adoc

Reviewer: I don't think we should expose "python-runtime" as part of this path as it's a very internal thing

Reviewer: (same for the other files)

Author: We use python runtime term in text too: `his chapter provides an overview of the available shallow embedding algorithms in the GDS Python runtime`
And we show PyTorch bind in the doc too
Do you want me to remove `python runtime` mentions from paths and text? Which term for change you suggest?

Reviewer: Yes, we should remove Python runtime from any user facing communication. Do we need another term at all? The only thing that's special about these algorithms for the user is that they can only run them through GDS Sessions

Author: I have removed "runtime" mentioning from the doc

= Knowledge graph embeddings

Knowledge Graph Embeddings (KGE) refer to a family of algorithms designed to learn low-dimensional representations of entities and relations within a knowledge graph.

> These embeddings are utilized for tasks including link prediction, entity classification, and entity clustering.

Reviewer: ```suggestion
These embeddings are utilized for tasks such as link prediction, entity classification, and entity clustering.
```

This chapter provides an overview of the available shallow embedding algorithms in the GDS Python runtime, such as `TransE` and `DistMult`.

* `TransE`

> ** https://proceedings.neurips.cc/paper_files/paper/2013/file/1cecc7a77928ca8133fa24680a88d2f9-Paper.pdf[Bordes, Antoine, et al. "Translating embeddings for modeling multi-relational data." Advances in neural information processing systems 26 (2013).^]

Reviewer: I think we have said in the past that we should not have links to papers as they may be broken. Instead just state the name of the authors and the paper

* `DistMult`
** https://arxiv.org/pdf/1412.6575[Yang, Bishan, et al. "Embedding entities and relations for learning and inference in knowledge bases." arXiv preprint arXiv:1412.6575 (2014).^]

A knowledge graph is a directed, multi-relational graph.
It consists of triples `(head, relation, tail)` where `head` and `tail` are entities and `relation` is the relationship between them.

Shallow models of KGE represented here are _inductive_ algorithms for computing node and relationship embeddings (low-dimensional representations) for knowledge graphs.
Shallow means that matrix lookups represent the entity and relation encoders.

> The algorithm is inductive, meaning that it can be trained on one graph and then applied to this exact graph or another graph with the same schema.

Reviewer: KGE models are not inductive. You get node embeddings as the output that only pertain the graph it was trained on

A KGE model can return the score for the triple which can be used to predict the missing tail entity for a given head and relation.
The embeddings can be used to perform various tasks, such as link prediction, entity classification, and entity clustering.

* https://arxiv.org/pdf/1706.02216.pdf[William L. Hamilton, Rex Ying, and Jure Leskovec. "Inductive Representation Learning on Large Graphs." 2018.^]
* https://arxiv.org/pdf/1911.10232.pdf[Amit Pande, Kai Ni and Venkataramani Kini. "SWAG: Item Recommendations using Convolutions on Weighted Graphs." 2019.^]

[[algorithms-embeddings-kge-considerations]]
== Scoring functions

For KGE algorithms, the scoring function is used to compute the score of a triple `(head, relation, tail)`.

> This score is used for computing loss during training and during prediction to rank the tail entities for a given head and relation.

Reviewer: I think we want to say "relationship type" instead of "relation", so as to adjust the terminology to Neo4j

Reviewer: Any comment on this?

Author: Thanks for noticing it again, I missed this point, now it's fixed

The scoring function is a function of the embeddings of the head, relation, and tail entities.

Let the embeddings of the head, relation, and tail entities be denoted by `h`, `r`, and `t` respectively.
To compute the score of the triple `(head, relation, tail)`, the scoring function takes the embeddings `h`, `r`, and `t` as input and computes a scalar score using the corresponding formula.

> === `TransE`

Reviewer: (nitpick) Our ascidoc convention is to have two blank lines before a new header

The formula to compute the score of a triple `(head, relation, tail)` using the embeddings `h`, `r`, and `t`:

image::python-runtime/distmult-formula.svg[width=400]

== Considerations
To effectively train KGE models, several considerations need to be taken into account, including the choice of loss function, sampling methods, and optimization strategies.

=== Stochastic Local Closed World Assumption (sLCWA)

Observed triplets in the knowledge graph are considered true.
The unobserved triplets can be treated differently based on the assumption made.

* Open World Assumption (OWA) assumes that all unobserved facts are unknown.

* Closed World Assumption (CWA) assumes that all unobserved facts are false.

* Local Closed World Assumption (LCWA) assumes that all observed facts are true.
All corrupted triplets, which are generated by replacing the head or tail entity of a positive triplet, are false.

* Stochastic Local Closed World Assumption (sLCWA) assumes that all observed facts are true.
Some corrupted triplets are false and some are true.
The number of corrupted triplets for each true triplet is set by the `negative_sampling_size` parameter.

> Knowledge graph embedding models are trained under the sLCWA assumption in current implementation.

Reviewer: Do we need to introduce all this terminology if we only use sLCWA?

Reviewer: Ok 👍

=== Loss function
The loss function is crucial for guiding the training process of KGE models.
It determines how the difference between predicted and actual values is calculated and minimized.
There are several loss functions that can be used for training KGE models, see below for more details.

==== Margin ranking loss

Margin ranking loss is a pairwise loss function that compares the scores based on the difference between the scores of a positive triple and a negative triple.

> When negative sampling size is more than one, the loss is computed for positive triple and each negative triple and the average loss is computed.

Reviewer: ```suggestion
When negative sampling size is more than one, the loss is computed for a positive triple and each of its negative triples, and the average loss is computed.
```

image::python-runtime/mrl.svg[width=300]
image::python-runtime/delta-value.svg[width=400]

==== Negative Sampling Self-Adversarial Loss

Negative Sampling Self-Adversarial Lossfootnote:[Sun, Zhiqing, et al. "Rotate: Knowledge graph embedding by relational rotation in complex space." arXiv preprint arXiv:1902.10197 (2019).] is a setwise loss function that compares the scores based on the difference between the scores of a positive triple and a set of negative triples.
`loss_function_kwargs` can be used to set the `adversarial_temperature` and `margin` parameters.

=== Optimizer

Any pytorch optimizer can be used for training the model.
To use non-default optimizer, specify the optimizer class name as a string in the `optimizer` parameter.
All optimizer parameters except `params` can be passed as `optimizer_kwargs`.

=== Negative sampling

> Loss function requires negative samples to compute the loss.

Reviewer: ```suggestion
The loss function requires negative samples to compute the loss.
```

The number of negative samples per positive sample is controlled by the `negative_sampling_size` parameter.
When `use_node_type_aware_sampler` is set to `True`, negative nodes are sampled with the same label as the corresponding positive node.
With or without node type awareness, the negative samples are sampled uniformly at random from the graph.

=== Learning rate scheduler

Any pytorch learning rate scheduler can be used for training the model.
To use non-default learning rate scheduler, specify the scheduler class name as a string in the `lr_scheduler` parameter.
All scheduler parameters except `optimizer` can be passed as `lr_scheduler_kwargs`.

=== Inner normalisation

In the original `TransE` paperfootnote:[Bordes, Antoine, et al. "Translating embeddings for modeling multi-relational data." Advances in neural information processing systems 26 (2013).]

> in `Algorithm 1`, line 5, the entity embeddings are normalized to have `Lp` norm of 1.

Reviewer: What is the value of `p` in this case?

Reviewer: Ok! Can the user set which `p` they want?

Author: Yes, it's the same as `p_norm` value, I'll add information about.

Reviewer: ok 👍

For some datasets, this normalization might not be beneficial.
To avoid this normalization, set `inner_norm` to `False`.

[[algorithms-embeddings-kge-syntax]]
== Syntax

[source,python]
----

> gds.model.transe.train(G,

Reviewer: As I've said before I would prefer something like `gds.kge.train(....)`

    num_epochs = 10,
)
----

.Parameters
[cols="1m,1m,1m,1", options="header"]
|====
| Parameter | Type | Default value | Description

| num_epochs
| int
| N/A
| Number of epochs for training (must be greater than 0)

| embedding_dimension
| int

> | 256

Reviewer: I don't think there should be a default value for embedding dimension. Better to force the user to specify as it depends so much on the data

| Dimensionality of the embeddings (must be greater than 0)

| epochs_per_checkpoint
| int
| max(num_epochs / 10, 1)
| Number of epochs between checkpoints (must be greater than or equal to 0)

| load_from_checkpoint
| Optional[tuple[str, int]]
| None
| Checkpoint to load from, specified as a tuple (path, epoch)

| split_ratios
| dict[str, float]
| {TRAIN=0.8, TEST=0.2}

> | Ratios for splitting the dataset into training and test sets

Reviewer: There's also validation right?

Reviewer: I think it's ok to not have it be default. Just specify a sum < 1.0 and the what remains should be allocated as validation set?

| scoring_function
| str
| "transe"
| Function used to score embeddings

| p_norm
| float
| 1.0
| Norm to use in TransE scoring function

| batch_size
| int

> | 512

Reviewer: I'm not sure it makes sense to have a default for this either? I think either we need to infer it intelligently from things like embedding size, GPU memory, etc, or not have a default. Wdyt? Same for `test_batch_size`

Reviewer: Ok! Yes, I agree it's good to let the user override this if they need to :)

| Size of the training batch (must be greater than 0)

| test_batch_size
| int
| 512
| Size of the test batch (must be greater than 0)

| optimizer
| str
| "adam"
| Optimizer to use for training

| optimizer_kwargs
| dict[str, Any]
| {lr=0.01, weight_decay=0.0005}

> | Additional arguments for the optimizer

Reviewer: ```suggestion
| Arguments for the optimizer
```

| lr_scheduler
| str
| ConstantLR
| Learning rate scheduler

| lr_scheduler_kwargs
| dict[str, Any]
| {factor=1, total_iters=1000}
| Additional arguments for the learning rate scheduler

| loss_function
| str
| MarginRanking
| Loss function to use for training

| loss_function_kwargs
| dict[str, Any]
| {margin=1.0, adversarial_temperature=1.0, gamma=20.0}
| Additional arguments for the loss function

| negative_sampling_size
| int
| 1
| Number of negative samples per positive sample

| use_node_type_aware_sampler
| bool
| False
| Whether to sample negative nodes with the same label as the corresponding positive node

| k_value
| int
| 10
| Value of k used in evaluation metrics

> | do_validation

Reviewer: What kind of validation is this referring to? For each epoch? Should not the `split_ratios` dictate whether validation should be done?

Author: Yes, if TRAIN + TEST < 1.0 we have to do validation, I'll put a clear statement about it.

| bool
| True
| Whether to perform validation

| do_test
| bool
| True
| Whether to perform testing

| filtered_metrics
| bool
| False

> | Whether to use filtered metrics during evaluation

Reviewer: Perhaps this needs to be explained further?

| epochs_per_val
| int
| 50
| Number of epochs between validations (must be greater than or equal to 0)

| disable_tqdm
| bool
| True

> | Whether to disable tqdm progress bars

Reviewer: TQDM is not something that is user facing right? I think what the user will see is something like MLFlow?

| inner_norm
| bool
| True
| Whether to apply normalization to embeddings

| init_bound
| Optional[float]
| None

> | Initial bound for embeddings (if any)

Reviewer: What does this mean?

Reviewer: ah ok!

| Additional arguments for the learning rate scheduler

| loss_function
| str
| MarginRanking
| Loss function to use for training

| loss_function_kwargs
| dict[str, Any]
| {margin=1.0, adversarial_temperature=1.0, gamma=20.0}
| Additional arguments for the loss function

| negative_sampling_size
| int
| 1
| Number of negative samples per positive sample

| use_node_type_aware_sampler
| bool
| False
| Whether to sample negative nodes with the same label as the corresponding positive node

| k_value
| int
| 10
| Value of k used in Hits@k evaluation metric

| do_validation
| bool
| True
| Whether to perform validation

| do_test
| bool
| True
| Whether to perform testing

| filtered_metrics
| bool
| False
| Whether to use filtered metrics during evaluation, see <<_filtered_metrics, filtered metrics>>

| epochs_per_val
| int
| 50
| Number of epochs between validations (must be greater than or equal to 0)

| inner_norm
| bool
| True
| Whether to apply normalization to embeddings, see <<_inner_normalisation, inner normalization>>

| init_bound
| Optional[float]
| None
| The value for the range [-init_bound; init_bound] of the uniform distribution used to initialize the embeddings.

> Xavier initializationfootnote:[Xavier Glorot, Yoshua Bengio "Understanding the difficulty of training deep feedforward neural networks" Proceedings of the Thirteenth International Conference on Artificial Intelligence and Statistics, PMLR 9:249-256, 2010]

Reviewer: Nice :)

= Knowledge graph embeddings

> 

Reviewer: Very nice intro!

Knowledge Graph Embeddings (KGE) refer to a family of algorithms designed to learn low-dimensional representations of entities and relations within a knowledge graph.
These embeddings are utilized for tasks such as link prediction, entity classification, and entity clustering.

This chapter provides an overview of the available shallow embedding algorithms in the GDS Python runtime, such as `TransE` and `DistMult`.

* `TransE`
** Bordes, Antoine, et al. "Translating embeddings for modeling multi-relational data." Advances in neural information processing systems 26 (2013).
* `DistMult`
** Yang, Bishan, et al. "Embedding entities and relations for learning and inference in knowledge bases." arXiv preprint arXiv:1412.6575 (2014)

A knowledge graph is a directed, multi-relational graph.
It consists of triples `(head, relation, tail)` where `head` and `tail` are entities and `relation` is the relationship between them.

> Shallow models of KGE represented here are _transductive_ algorithms for computing node and relationship embeddings (low-dimensional representations) for knowledge graphs.

Reviewer: ```suggestion
Shallow models of KGE represented here are _transductive_ algorithms for computing node and relationship embeddings (low-dimensional representations) for knowledge graphs.
```
"relationship" -> "relation" or "relationship type"?

> Shallow means that matrix lookups represent the entity and relation encoders.

Author: Term "shallow" doesn't relate to the method used to train the model. It is about the fact that the model on inference stage returns embedding as result of lookup operation, without any additional computing/processing.
Ref: https://arxiv.org/pdf/1709.05584
<img width="1214" alt="image" src="https://github.com/neo4j/graph-data-science-client/assets/50139720/4c8f8409-621a-4df7-8538-dd0cf1e74b59">

> Shallow means that matrix lookups represent the entity and relation encoders.

Reviewer: I don't disagree with this. But nevertheless I think the word shallow is used in relation to its opposite "deep" for deep learning, such as GNNs

The algorithm is inductive, meaning that it can be trained on one graph and then applied to this exact graph or another graph with the same schema.

> A KGE model can return the score for the triple which can be used to predict the missing tail entity for a given head and relationship type.

Author: Currently we compute scores for tails. Head can be predicted as part of additional prediction function or as tail for inverse relationship which is not in the scope now.

> A KGE model can return the score for the triple which can be used to predict the missing tail entity for a given head and relationship type.

Reviewer: Ok. It's a fairly common use-case but I guess we could add it later

The embeddings can be used to perform various tasks, such as link prediction, entity classification, and entity clustering.

[[algorithms-embeddings-kge-considerations]]
== Scoring functions

> For KGE algorithms, the scoring function is used to compute the score of a triple `(head, relation, tail)`.

Author: Repetition is the mother of learning 😄

> For KGE algorithms, the scoring function is used to compute the score of a triple `(head, relation, tail)`.

Reviewer: OK 👍

This score is used for computing loss during training and during prediction to rank the tail entities for a given head and relation.

The scoring function is a function of the embeddings of the head, relation, and tail entities.

Let the embeddings of the head, relation, and tail entities be denoted by `h`, `r`, and `t` respectively.
To compute the score of the triple `(head, relation, tail)`, the scoring function takes the embeddings `h`, `r`, and `t` as input and computes a scalar score using the corresponding formula.

=== `TransE`

> `TransE` is a translational distance interaction model.

Author: Yes, ref: https://arxiv.org/pdf/2006.13365 Chapter 3.1

> `TransE` is a translational distance interaction model.

Reviewer: Interesting, I've never heard that terminology anywhere else in the literature. Are we sure this is not just something that these particular authors invented for this paper?

It determines how the difference between predicted and actual values is calculated and minimized.
There are several loss functions that can be used for training KGE models, see below for more details.

==== Margin ranking loss

Margin ranking loss is a pairwise loss function that compares the scores based on the difference between the scores of a positive triple and a negative triple.
When negative sampling size is more than one, the loss is computed for a positive triple and each of its negative triples, and the average loss is computed.

image::python-runtime/mrl.svg[width=300]
image::python-runtime/delta-value.svg[width=400]

==== Negative Sampling Self-Adversarial Loss

Negative Sampling Self-Adversarial Lossfootnote:[Sun, Zhiqing, et al. "Rotate: Knowledge graph embedding by relational rotation in complex space." arXiv preprint arXiv:1902.10197 (2019).] is a setwise loss function that compares the scores based on the difference between the scores of a positive triple and a set of negative triples.
`loss_function_kwargs` can be used to set the `adversarial_temperature` and `margin` parameters.

=== Optimizer

> Any pytorch optimizer can be used for training the model.

Reviewer: I'm worried about this. It really depends on the PyTorch version. I think it's quite dangerous to expose this sort of thing

Reviewer: Better, but I think we should probably explicitly list exactly the optimizers we support. Similar for schedulers of course. And what happens if these optimizers changes name between PyTorch versions...

In general I'm not super fond of us exposing the implementation details like this. But it's something Gian wants I guess

Reviewer: I think we have to do that, or state all names of the optimizers/schedulers, and exactly what parameters they can take. Let's think about it

=== Learning rate scheduler

Any pytorch learning rate scheduler can be used for training the model.
To use non-default learning rate scheduler, specify the scheduler class name as a string in the `lr_scheduler` parameter.
All scheduler parameters except `optimizer` can be passed as `lr_scheduler_kwargs`.

=== Inner normalisation

In the original `TransE` paperfootnote:[Bordes, Antoine, et al. "Translating embeddings for modeling multi-relational data." Advances in neural information processing systems 26 (2013).]
in `Algorithm 1`, line 5, the entity embeddings are normalized to have `Lp` norm of 1.
Value of `p` is set by the `p_norm` parameter.
For some datasets, this normalization might not be beneficial.
To avoid this normalization, set `inner_norm` to `False`.

=== Filtered metrics

When we evaluate (compute metrics) the model, we compute scores for all possible triples with the same head or tail and relationship type as the test triple.
Ranking the test triple among other triples is used for computing metrics, such as Mean Rank, Mean Reciprocal Rank, and Hit@k.

When `filtered_metrics` is set to `False`, the ranking is done among all possible triples.

> When `filtered_metrics` is set to `True`, the ranking is done among only the triples that are not present in the training set.

Reviewer: Nice explanation :)

[[algorithms-embeddings-kge-syntax]]
== Syntax

[source, python, role=no-test]
----

> gds.model.train(G,

Reviewer: I don't think that `model` is sufficiently clear, as there are several types of models in the library. Did you consider `kge` or something instead? Or perhaps something else

Author: Yes, I think we have to mention `kge`.

    num_epochs = 10,
    embedding_dimension = 100,
)
----

.Parameters
[cols="1m,1m,1m,1", options="header"]
|====
| Parameter | Type | Default value | Description

| num_epochs
| int
| N/A
| Number of epochs for training (must be greater than 0)

| embedding_dimension
| int
| N/A
| Dimensionality of the embeddings (must be greater than 0)

| epochs_per_checkpoint
| int
| max(num_epochs / 10, 1)
| Number of epochs between checkpoints (must be greater than or equal to 0)

| load_from_checkpoint
| Optional[tuple[str, int]]
| None

> | Checkpoint to load from, specified as a tuple (path, epoch)

Reviewer: I forget how it works.
```suggestion
| Checkpoint to load from, specified as a tuple (modelname, epoch)
```
I think it's just the model name right?

Author: Yes

| split_ratios
| dict[str, float]
| {TRAIN=0.8, TEST=0.2}
| Ratios for splitting the dataset into training and test sets.

> When sum of values is less than 1, the remaining data is used for validation.

Reviewer: ```suggestion
When the sum of the ratios is less than 1.0, the remaining examples are used for validation.
```

> Validation size can be set explicitly with the key `VALID`.

Reviewer: ```suggestion
The validation set ratio can be set explicitly with the key `VALID`.
```

> When all three keys are present, the sum of values must be equal to 1.

Reviewer: ```suggestion
When all three keys are present, the sum of values must be equal to 1.0.
```

| scoring_function
| str
| "transe"

> | Function used to score embeddings

Reviewer: ```suggestion
| Function used to score embeddings of triples
```
Just to clarify :)

| p_norm
| float
| 1.0
| Norm to use in TransE scoring function

| batch_size
| int
| 512
| Size of the training batch (must be greater than 0)

| test_batch_size
| int
| 512
| Size of the test batch (must be greater than 0)

| optimizer
| str
| "adam"
| Optimizer to use for training

| optimizer_kwargs
| dict[str, Any]
| {lr=0.01, weight_decay=0.0005}
| Arguments for the optimizer

| lr_scheduler
| str
| ConstantLR
| Learning rate scheduler

| lr_scheduler_kwargs
| dict[str, Any]
| {factor=1, total_iters=1000}
| Additional arguments for the learning rate scheduler

| loss_function
| str
| MarginRanking
| Loss function to use for training

| loss_function_kwargs
| dict[str, Any]
| {margin=1.0, adversarial_temperature=1.0, gamma=20.0}
| Additional arguments for the loss function

| negative_sampling_size
| int
| 1
| Number of negative samples per positive sample

| use_node_type_aware_sampler
| bool
| False
| Whether to sample negative nodes with the same label as the corresponding positive node

| k_value
| int
| 10

> | Value of k used in evaluation metrics

Reviewer: ```suggestion
| Value of k used in Hits@k evaluation metric
```
Or are there more?

Author: No, nothing more

| do_validation
| bool
| True
| Whether to perform validation

| do_test
| bool
| True
| Whether to perform testing

| filtered_metrics
| bool
| False
| Whether to use filtered metrics during evaluation, see <<_filtered_metrics, filtered metrics>>

| epochs_per_val
| int
| 50
| Number of epochs between validations (must be greater than or equal to 0)

| inner_norm
| bool
| True
| Whether to apply normalization to embeddings, see <<_inner_normalisation, inner normalization>>

| init_bound
| Optional[float]
| None

> | The value for the range [-bound; bound] of the uniform distribution used to initialize the embeddings. Xavier's initialization is used if None.

Reviewer: ```suggestion
| The value for the range [-init_bound; init_bound] of the uniform distribution used to initialize the embeddings. Xavier initialization is used if None.
```
Also, should we explain or refer to somewhere for more on Xavier initialization?

Author: Added footnote

|====

.Results
[opts="header",cols="2m,1,6"]
|===
| Name              | Type    | Description
| modelInfo         | Map     | Details of the trained model.
| configuration     | Map     | The configuration used to run the procedure.

> | trainMillis       | Integer | Milliseconds to train the model.

Reviewer: I don't think this is exactly what we have atm? We have some `metadata` field atm I think. Also, Should we still use the Cypher data types when this is just going to be all Python?

Author: I removed this chapter, until we have the right one
